### DISCUSSION TITLE
How do I find the location of an installed binary from a package?

### DISCUSSION THREAD

**Discussion by lasagnaphil** [xmake-io/xmake#2219] [Q&A] [answered]
I am currently writing an application that needs some code generation before C++ compliation. More specifically, I'm using a tool called `sokol-shdc` which converts GLSL shaders to header files. sokol-shdc is provided by a custom package called sokol-tools-bin, which the description file is the following (just copies an executable from a git repository):

```lua
package("sokol-tools-bin")
    set_kind("library", {binary = true})
    add_urls("https://github.com/floooh/sokol-tools-bin.git")
    on_install("windows", function (package)
        os.cp("bin/win32/sokol-shdc.exe", package:installdir("bin"))
    end)
```

Now I'm trying to create a custom rule to invoke sokol-shdc for all files with the `*.glsl` extension.  I need a nice and easy way to find where "sokol-shdc.exe" is installed in my system, so naturally I use `find_package()` to get the install directory of the `sokol-tools-bin` package.

```lua
rule("sokol_shaders")
	set_extensions(".glsl")
	on_load(function (target)
		target:add("includedirs", target:autogendir())
	end)
	before_buildcmd_file(function (target, batchcmds, shaderfile, opt)
		import("lib.detect.find_package")

		local headerdir = target:autogendir()
		local name = path.basename(shaderfile)
		local ext = path.extension(shaderfile)
		local headerfile = path.join(headerdir, string.format("%s.%s.h", name, ext))

		local package = find_package("sokol-tools-bin")
		local binary = path.join(package:installdir("bin"), "sokol-shdc.exe")

		batchcmds:show_progress(opt.progress, "${color.build.object}generating headers for %s", shaderfile)
		batchcmds:vrunv(binary, "--input", shaderfile, "--output", headerfile, "--slang", "glsl330:hlsl5:metal_macos")
		batchcmds:add_depfiles(shaderfile)
		batchcmds:set_depmtime(os.mtime(headerfile))
		batchcmds:set_depcache(target:dependfile(headerfile))
	end)
rule_end()
```

Then I apply this rule to my target:

```lua
target("renderer_test")
	add_rules("sokol_shaders")
	set_kind("binary")
	add_files("renderer/*.cpp")
	add_files("renderer/shaders/*.glsl")
	add_includedirs("renderer")
	add_packages("opengl", "sokol", "sokol-tools-bin")
```

The problem is, `find_package("sokol-tools-bin")` returns nil and xmake gives out an error. I'm suspecting the reason for this is that the packages for the target `renderer_test` are added later than `before_buildcmd_file()` callback is called. Is there a better way to find the path of an installed package?

**Comment by waruqi** [comment, marked as answer]
why do not use `bin2c` or `glsl2spv` rules?

https://xmake.io/#/manual/custom_rule?id=utilsbin2c
https://xmake.io/#/manual/custom_rule?id=utilsglsl2spv

or you can refer https://github.com/xmake-io/xmake/blob/master/xmake/rules/utils/glsl2spv/xmake.lua

**Reply by lasagnaphil** [reply]
`sokol-shdc` does a bit more than what `bin2c` and `glsl2spv` does. It converts a GLSL-like shader code to whatever format you need depending on the rendering backend (GLSL, HLSL, Metal Shading Language, ...), and is tailor-made to work with sokol (https://github.com/floooh/sokol). 

The code for `glsl2spv` that you've linked also has the same problem as what I have currently: it assumes that glslangValidator (or glslc) is installed on the system. Ideally for user convenience, it would be better if a xmake package named `glslangValidator-bin` provided the executable, and the `glsl2spv` rule can use the binary from that package. In that way the user doesn't need to install `glslangValidator` separately from xmake.

**Reply by waruqi** [reply]
> `sokol-shdc` does a bit more than what `bin2c` and `glsl2spv` does. It converts a GLSL-like shader code to whatever format you need depending on the rendering backend (GLSL, HLSL, Metal Shading Language, ...), and is tailor-made to work with sokol (https://github.com/floooh/sokol).

so you can refer https://github.com/xmake-io/xmake/blob/master/xmake/rules/utils/glsl2spv/xmake.lua and improve your rule.



> 
> The code for `glsl2spv` that you've linked also has the same problem as what I have currently: it assumes that glslangValidator (or glslc) is installed on the system. Ideally for user convenience, it would be better if a xmake package named `glslangValidator-bin` provided the executable, and the `glsl2spv` rule can use the binary from that package. In that way the user doesn't need to install `glslangValidator` separately from xmake.

no, we will add glslang package, it will add glslangValidator bin path to `$PATH`.

```lua
add_requires("glslang", {configs = {binaryonly = true}})
target("test")
     add_rules("utils.glsl2spv")
     add_packages("glslang")
```

glslang package will find glslangValidator on system and add bin path, it will install it automatically if not found.

**Reply by lasagnaphil** [reply]
Is there a way to use the `glslangValidator` binary from that package without adding to the $PATH? I would like to keep the environment variables as clean as possible, since multiple projects can possibly use different versions of the `glslangValidator` executable.

**Reply by waruqi** [reply]
> Is there a way to use the `glslangValidator` binary from that package without adding to the $PATH? I would like to keep the environment variables as clean as possible, since multiple projects can possibly use different versions of the `glslangValidator` executable.

`add_packages("glslang")` just add to $PATH in given target. only for their rules environment (on_xxx, before_xxx, after_xxx), so it's very clean.

you can add different version of glslang package to the different targets.

**Reply by waruqi** [reply]
```lua
add_requires("glslang 1.0", {configs = {binaryonly = true}, alias = "glslang1x"})
add_requires("glslang 2.0", {configs = {binaryonly = true}, alias = "glslang2x"})

target("test")
     add_rules("utils.glsl2spv")
     add_packages("glslang1x")

target("test2")
     add_rules("utils.glsl2spv")
     add_packages("glslang2x")
```

**Reply by lasagnaphil** [reply]
Understood. I think I would be able to do what I want by following the code for glsl2spv. Thanks for all the hard work!
